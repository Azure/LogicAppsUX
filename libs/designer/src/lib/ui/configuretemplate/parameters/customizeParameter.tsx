import { TemplatesSection, type TemplatesSectionItem } from '@microsoft/designer-ui';
import { useResourceStrings } from '../resources';
import { useMemo } from 'react';
import { getResourceNameFromId, type Template } from '@microsoft/logic-apps-shared';
import { useSelector } from 'react-redux';
import type { RootState } from '../../../core/state/templates/store';

export const CustomizeParameter = ({
  parameterDefinition,
  setParameterDefinition,
}: {
  parameterDefinition: Template.ParameterDefinition;
  setParameterDefinition: (parameterDefinition: Template.ParameterDefinition) => void;
}) => {
  const resourceStrings = useResourceStrings();

  const { isAccelerator } = useSelector((state: RootState) => ({
    isAccelerator: Object.keys(state.template.workflows).length > 1,
  }));

  const detailsSectionItems: TemplatesSectionItem[] = useMemo(() => {
    const baseItems: TemplatesSectionItem[] = [
      {
        label: resourceStrings.ParameterName,
        value: parameterDefinition.name || '',
        type: 'text',
      },
      {
        label: resourceStrings.Type,
        value: parameterDefinition.type || '',
        type: 'text',
      },
      {
        label: resourceStrings.ParameterDisplayName,
        value: parameterDefinition.displayName || '',
        type: 'textfield',
        required: true,
        onChange: (value: string) => {
          setParameterDefinition({
            ...parameterDefinition,
            displayName: value,
          });
        },
      },
      {
        label: resourceStrings.DefaultValue,
        value: parameterDefinition.default || '',
        type: 'textfield',
        onChange: (value: string) => {
          setParameterDefinition({
            ...parameterDefinition,
            default: value,
          });
        },
      },
      {
        label: resourceStrings.Description,
        value: parameterDefinition.description || '',
        type: 'textarea',
        onChange: (value: string) => {
          setParameterDefinition({
            ...parameterDefinition,
            description: value,
          });
        },
      },
      {
        label: resourceStrings.RequiredField,
        value: parameterDefinition.required ?? false,
        type: 'switch',
        onChange: (value: boolean) => {
          setParameterDefinition({
            ...parameterDefinition,
            required: value,
          });
        },
      },
    ];
    if (isAccelerator && parameterDefinition.associatedWorkflows) {
      baseItems.push({
        label: resourceStrings.AssociatedWorkflows,
        value: formatAssociatedWorklows(parameterDefinition.associatedWorkflows) || '',
        type: 'text',
      });
    }
    return baseItems;
  }, [resourceStrings, parameterDefinition, setParameterDefinition, isAccelerator]);

  return (
    <div>
      <TemplatesSection title={resourceStrings.Details} titleHtmlFor={'detailsSectionLabel'} items={detailsSectionItems} />
    </div>
  );
};

const formatAssociatedWorklows = (associatedWorkflows: string[]) => {
  return associatedWorkflows.map((workflow) => getResourceNameFromId(workflow)).join(', ');
};
