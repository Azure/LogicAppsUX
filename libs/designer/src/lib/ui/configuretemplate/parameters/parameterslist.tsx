import {
  Button,
  Link,
  Table,
  TableBody,
  TableCell,
  TableCellLayout,
  TableHeader,
  TableHeaderCell,
  TableRow,
  Text,
} from '@fluentui/react-components';
import type { AppDispatch, RootState } from '../../../core/state/templates/store';
import { useDispatch, useSelector } from 'react-redux';
import { useResourceStrings } from '../resources';
import { useCallback, useMemo } from 'react';
import { useIntl } from 'react-intl';
import { getResourceNameFromId } from '@microsoft/logic-apps-shared';
import { openPanelView, TemplatePanelView } from '../../../core/state/templates/panelSlice';
import { MoreHorizontal16Filled } from '@fluentui/react-icons';
import { CustomizeParameterPanel } from '../panels/customizeParameterPanel/customizeParameterPanel';
import { DescriptionWithLink, ErrorBar } from '../common';
import { mergeStyles } from '@fluentui/react';
import { formatNameWithIdentifierToDisplay } from '../../../core/configuretemplate/utils/helper';

const columnTextStyle: React.CSSProperties = {
  display: '-webkit-box',
  WebkitLineClamp: 2,
  WebkitBoxOrient: 'vertical',
  overflow: 'hidden',
  textOverflow: 'ellipsis',
  wordBreak: 'break-word',
  lineBreak: 'anywhere',
};

export const TemplateParametersList = () => {
  const intl = useIntl();
  const dispatch = useDispatch<AppDispatch>();

  const intlText = {
    AriaLabel: intl.formatMessage({
      defaultMessage: 'List of parameters in the template',
      id: 'u2z3kg',
      description: 'The aria label for the parameters table',
    }),
    Description: intl.formatMessage({
      defaultMessage:
        'Customize each parameter to tailor this template to your needs. These values help configure how your workflows run. You can save your progress anytime and return later to finish, but all fields must be completed for the template to work.',
      id: 'nCjxEh',
      description: 'The description for the parameters tab',
    }),
    ErrorTitle: intl.formatMessage({
      defaultMessage: 'Validation failed for parameters: ',
      id: 'MQ0ODD',
      description: 'The error title for the parameters tab',
    }),
  };

  const { parameterDefinitions, currentPanelView, workflowsInTemplate, parameterErrors } = useSelector((state: RootState) => ({
    parameterDefinitions: state.template.parameterDefinitions,
    currentPanelView: state.panel.currentPanelView,
    workflowsInTemplate: state.template.workflows,
    parameterErrors: state.template.errors.parameters,
  }));

  const formattedParameterErrorIds = useMemo(() => {
    return Object.entries(parameterErrors)
      .filter(([_id, error]) => error)
      .map(([id]) => formatNameWithIdentifierToDisplay(id));
  }, [parameterErrors]);

  const isAccelerator = Object.keys(workflowsInTemplate).length > 1;
  const resourceStrings = useResourceStrings();

  const columns = useMemo(() => {
    const baseColumn = [
      { columnKey: 'name', label: resourceStrings.Name },
      { columnKey: 'displayName', label: resourceStrings.DisplayName },
      { columnKey: 'type', label: resourceStrings.Type },
      { columnKey: 'default', label: resourceStrings.DefaultValue },
      // { columnKey: 'allowedValues', label: resourceStrings.AllowedValues },  //TODO: revisit allowedValues
    ];
    const column2 = [
      { columnKey: 'description', label: resourceStrings.Description },
      { columnKey: 'required', label: resourceStrings.Required },
    ];
    if (isAccelerator) {
      return [...baseColumn, { columnKey: 'associatedWorkflows', label: resourceStrings.AssociatedWorkflows }, ...column2];
    }
    return [...baseColumn, ...column2];
  }, [isAccelerator, resourceStrings]);

  const items = useMemo(
    () =>
      Object.values(parameterDefinitions)?.map((parameter) => ({
        name: parameter?.name ?? resourceStrings.Placeholder,
        displayName: parameter?.displayName ?? resourceStrings.Placeholder,
        type: parameter.type,
        default: parameter?.default ?? resourceStrings.Placeholder,
        // allowedValues:
        //   parameter?.allowedValues?.map((allowedValue) => `${allowedValue.displayName} ${allowedValue.value}`)?.join(', ') ??
        //   resourceStrings.Placeholder,
        associatedWorkflows: parameter?.associatedWorkflows?.join(', ') ?? resourceStrings.Placeholder,
        description: parameter?.description ?? resourceStrings.Placeholder,
        required: parameter?.required ?? false,
      })) ?? [],
    [parameterDefinitions, resourceStrings]
  );

  const handleSelectParameter = useCallback(
    (parameterId: string) => {
      dispatch(openPanelView({ panelView: TemplatePanelView.CustomizeParameter, selectedTabId: parameterId }));
    },
    [dispatch]
  );

  if (Object.keys(parameterDefinitions).length === 0) {
    return (
      <div className="msla-templates-wizard-tab-content" style={{ overflowX: 'auto', paddingTop: '12px' }}>
        <Text>{resourceStrings.NoParameterInTemplate}</Text>
      </div>
    );
  }

  return (
    <div className="msla-templates-wizard-tab-content" style={{ overflowX: 'auto', paddingTop: '12px' }}>
      {currentPanelView === TemplatePanelView.CustomizeParameter && <CustomizeParameterPanel />}
      <DescriptionWithLink
        text={intlText.Description}
        linkText={resourceStrings.LearnMore}
        linkUrl="https://go.microsoft.com/fwlink/?linkid=2321714"
        className={mergeStyles({ marginLeft: '-10px', width: '70%' })}
      />

      {formattedParameterErrorIds.length ? (
        <ErrorBar title={intlText.ErrorTitle} errorMessage={formattedParameterErrorIds.join(', ')} styles={{ marginLeft: '-10px' }} />
      ) : null}

      <Table aria-label={intlText.AriaLabel} size="small" style={{ width: '100%' }}>
        <TableHeader>
          <TableRow>
            {columns.map((column) => (
              <TableHeaderCell key={column.columnKey}>
                <Text weight="semibold">{column.label}</Text>
              </TableHeaderCell>
            ))}
          </TableRow>
        </TableHeader>
        <TableBody>
          {items.map((item) => (
            <TableRow key={item.name}>
              <TableCell>
                <TableCellLayout
                  style={{
                    overflow: 'hidden',
                  }}
                >
                  <Link
                    style={columnTextStyle}
                    as="button"
                    onClick={() => {
                      handleSelectParameter(item.name);
                    }}
                  >
                    {formatNameWithIdentifierToDisplay(item.name)}
                  </Link>
                </TableCellLayout>
              </TableCell>
              <TableCell>
                <TableCellLayout>
                  <Text style={columnTextStyle}>{item.displayName}</Text>
                </TableCellLayout>
              </TableCell>
              <TableCell>
                <TableCellLayout>
                  <Text style={columnTextStyle}>{item.type}</Text>
                </TableCellLayout>
              </TableCell>
              <TableCell>
                <TableCellLayout>
                  <Text style={columnTextStyle}>{item.default}</Text>
                </TableCellLayout>
              </TableCell>
              {/* <TableCell>
                <TableCellLayout>{item.allowedValues}</TableCellLayout>
              </TableCell> */}
              {isAccelerator && (
                <TableCell>
                  <TableCellLayout>
                    <Text style={columnTextStyle}>{getResourceNameFromId(item.associatedWorkflows)}</Text>
                  </TableCellLayout>
                </TableCell>
              )}
              <TableCell>
                <TableCellLayout>
                  <Text style={columnTextStyle}>{item.description}</Text>
                </TableCellLayout>
              </TableCell>
              <TableCell>
                <TableCellLayout>{item.required ? resourceStrings.RequiredOn : resourceStrings.RequiredOff}</TableCellLayout>
              </TableCell>
              <TableCell>
                <TableCellLayout>
                  <Button icon={<MoreHorizontal16Filled />} appearance="subtle" onClick={() => handleSelectParameter(item.name)} />
                </TableCellLayout>
              </TableCell>
            </TableRow>
          ))}
        </TableBody>
      </Table>
    </div>
  );
};
