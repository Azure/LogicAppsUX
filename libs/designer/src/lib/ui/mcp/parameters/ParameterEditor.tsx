import {
  ArrayEditor,
  AuthenticationEditor,
  type AuthenticationEditorOptions,
  type ChangeState,
  Combobox,
  DictionaryEditor,
  DropdownEditor,
  DynamicLoadStatus,
  FilePickerEditor,
  getDropdownOptionsFromOptions,
  HTMLEditor,
  mergeClasses,
  SchemaEditor,
  StringEditor,
  TableEditor,
} from '@microsoft/designer-ui';
import { getPropertyValue, replaceWhiteSpaceWithUnderscore, type ParameterInfo, type ValueSegment } from '@microsoft/logic-apps-shared';
import { useEditOperationStyles } from './styles';
import { useCallback } from 'react';
import {
  getDisplayValueFromPickerSelectedItem,
  getValueFromPickerSelectedItem,
  loadDynamicTreeItemsForParameter,
  loadDynamicValuesForParameter,
  parameterValueToString,
  shouldEncodeParameterValueForOperationBasedOnMetadata,
} from '../../../core/utils/parameters/helper';
import { useDispatch, useSelector } from 'react-redux';
import type { AppDispatch, RootState } from '../../../core/state/mcp/store';
import constants from '../../../common/constants';
import { useConnectorName } from '../../../core/state/selectors/actionMetadataSelector';

interface ParameterEditorProps {
  operationId: string;
  groupId: string;
  parameter: ParameterInfo;
  onParameterVisibilityUpdate: () => void;
  onParameterValueChange: (newState: ChangeState) => void;
}

const mcpEditorsPlugin = {
  tokens: false,
};

export const ParameterEditor = ({
  operationId,
  groupId,
  parameter,
  onParameterVisibilityUpdate,
  onParameterValueChange,
}: ParameterEditorProps) => {
  const styles = useEditOperationStyles();
  const dispatch = useDispatch<AppDispatch>();
  const { operationInfo, reference, nodeInputs, dependencies } = useSelector((state: RootState) => ({
    operationInfo: state.operations.operationInfo[operationId],
    reference: state.connection.connectionReferences[state.connection.connectionsMapping[operationId] ?? ''],
    nodeInputs: state.operations.inputParameters[operationId],
    dependencies: state.operations.dependencies[operationId],
  }));
  const displayNameResult = useConnectorName(operationInfo);

  const onComboboxMenuOpen = useCallback((): void => {
    if (parameter?.dynamicData?.status === DynamicLoadStatus.FAILED || parameter?.dynamicData?.status === DynamicLoadStatus.NOTSTARTED) {
      loadDynamicValuesForParameter(
        operationId,
        groupId,
        parameter?.id,
        operationInfo,
        reference,
        nodeInputs,
        dependencies,
        true /* showErrorWhenNotReady */,
        dispatch,
        {},
        {}
      );
    }
  }, [dependencies, dispatch, groupId, nodeInputs, operationId, operationInfo, parameter?.dynamicData?.status, parameter?.id, reference]);

  const getPickerCallbacks = () => ({
    getFileSourceName: () => displayNameResult.result,
    getDisplayValueFromSelectedItem: (selectedItem: any) =>
      getDisplayValueFromPickerSelectedItem(selectedItem, parameter as ParameterInfo, dependencies),
    getValueFromSelectedItem: (selectedItem: any) => getValueFromPickerSelectedItem(selectedItem, parameter as ParameterInfo, dependencies),
    onFolderNavigation: (selectedItem: any | undefined): void => {
      loadDynamicTreeItemsForParameter(
        operationId,
        groupId,
        parameter?.id ?? '',
        selectedItem,
        operationInfo,
        reference,
        nodeInputs,
        dependencies,
        true /* showErrorWhenNotReady */,
        dispatch,
        /* idReplacements */ {},
        /* workflowParameters */ {}
      );
    },
  });

  const onCastParameter = (value: ValueSegment[]): string => {
    return (
      parameterValueToString(
        {
          ...parameter,
          value,
        } as ParameterInfo,
        /* isDefinitionValue */ false,
        /* idReplacements */ {},
        shouldEncodeParameterValueForOperationBasedOnMetadata(operationInfo)
      ) ?? ''
    );
  };

  const dropdownOptions = getDropdownOptionsFromOptions(parameter.editorOptions);
  const labelForAutomationId = replaceWhiteSpaceWithUnderscore(parameter.label);
  const editorClassName = mergeClasses('msla-setting-token-editor-container', styles.parameterEditor);

  switch (parameter.editor?.toLowerCase()) {
    case constants.EDITOR.ARRAY:
      return (
        <ArrayEditor
          className={editorClassName}
          labelId={labelForAutomationId}
          arrayType={parameter.editorViewModel.arrayType}
          initialMode={parameter.editorOptions?.initialMode}
          label={parameter.label}
          initialValue={parameter.value}
          basePlugins={mcpEditorsPlugin}
          itemSchema={parameter.editorViewModel.itemSchema}
          onChange={onParameterValueChange}
          options={dropdownOptions}
          onMenuOpen={onComboboxMenuOpen}
          castParameter={onCastParameter}
          dataAutomationId={`msla-setting-token-editor-combobox-${labelForAutomationId}`}
        />
      );

    case constants.EDITOR.AUTHENTICATION:
      return (
        <AuthenticationEditor
          className={editorClassName}
          labelId={labelForAutomationId}
          initialValue={parameter.value}
          options={parameter.editorOptions as AuthenticationEditorOptions}
          type={parameter.editorViewModel.type}
          authenticationValue={parameter.editorViewModel.authenticationValue}
          onChange={onParameterValueChange}
          basePlugins={mcpEditorsPlugin}
        />
      );

    case constants.EDITOR.DICTIONARY:
      return (
        <DictionaryEditor
          className={editorClassName}
          labelId={labelForAutomationId}
          label={parameter.label}
          basePlugins={mcpEditorsPlugin}
          readonly={parameter.editorOptions?.readOnly}
          initialValue={parameter.value}
          initialItems={parameter.editorViewModel.items}
          valueType={parameter.editorOptions?.valueType}
          onChange={onParameterValueChange}
          dataAutomationId={`msla-setting-token-editor-dictionaryeditor-${labelForAutomationId}`}
        />
      );

    case constants.EDITOR.DROPDOWN:
      return (
        <DropdownEditor
          className={styles.parameterEditor}
          label={parameter.label}
          initialValue={parameter.value}
          options={dropdownOptions.map((option: any, index: number) => ({
            key: index.toString(),
            ...option,
          }))}
          multiSelect={!!getPropertyValue(parameter.editorOptions, 'multiSelect')}
          serialization={parameter.editorOptions?.serialization}
          onChange={onParameterValueChange}
          dataAutomationId={`msla-setting-token-editor-dropdowneditor-${labelForAutomationId}`}
        />
      );

    case constants.EDITOR.COMBOBOX:
      return (
        <Combobox
          className={editorClassName}
          labelId={labelForAutomationId}
          label={parameter.label}
          basePlugins={mcpEditorsPlugin}
          initialValue={parameter.value}
          options={dropdownOptions}
          useOption={true}
          isLoading={parameter.dynamicData?.status === DynamicLoadStatus.LOADING}
          errorDetails={parameter.dynamicData?.error ? { message: parameter.dynamicData.error.message } : undefined}
          onChange={onParameterValueChange}
          onMenuOpen={onComboboxMenuOpen}
          multiSelect={getPropertyValue(parameter.editorOptions, 'multiSelect')}
          serialization={parameter.editorOptions?.serialization}
          dataAutomationId={`msla-setting-token-editor-combobox-${labelForAutomationId}`}
        />
      );

    case constants.EDITOR.FILEPICKER:
      return (
        <FilePickerEditor
          className={editorClassName}
          labelId={labelForAutomationId}
          basePlugins={{ ...mcpEditorsPlugin, clearEditor: true }}
          initialValue={parameter.value}
          displayValue={parameter.editorViewModel.displayValue}
          type={parameter.editorOptions?.pickerType}
          items={parameter.editorOptions?.items}
          fileFilters={parameter.editorOptions?.fileFilters}
          pickerCallbacks={getPickerCallbacks()}
          isLoading={parameter.dynamicData?.status === DynamicLoadStatus.LOADING}
          errorDetails={parameter.dynamicData?.error ? { message: parameter.dynamicData.error.message } : undefined}
          editorBlur={onParameterValueChange}
          dataAutomationId={`msla-setting-token-editor-filepickereditor-${labelForAutomationId}`}
        />
      );

    case constants.EDITOR.HTML:
      return (
        <HTMLEditor
          className={styles.parameterEditor}
          labelId={labelForAutomationId}
          initialValue={parameter.value}
          basePlugins={mcpEditorsPlugin}
          onChange={onParameterValueChange}
          valueType={constants.SWAGGER.TYPE.ANY}
          dataAutomationId={`msla-setting-token-editor-htmleditor-${labelForAutomationId}`}
        />
      );

    case constants.EDITOR.SCHEMA:
      return (
        <SchemaEditor
          className={styles.parameterEditor}
          label={parameter.label}
          initialValue={parameter.value}
          onChange={onParameterValueChange}
        />
      );

    case constants.EDITOR.TABLE:
      return (
        <TableEditor
          className={styles.parameterEditor}
          labelId={labelForAutomationId}
          initialValue={parameter.value}
          initialItems={parameter.editorViewModel.items}
          columnMode={parameter.editorViewModel.columnMode}
          columns={parameter.editorOptions?.columns?.count}
          titles={parameter.editorOptions?.columns?.titles}
          keys={parameter.editorOptions?.columns?.keys}
          types={parameter.editorOptions?.columns?.types}
          onChange={onParameterValueChange}
          dataAutomationId={`msla-setting-token-editor-tableditor-${labelForAutomationId}`}
        />
      );

    default:
      return (
        <StringEditor
          className={editorClassName}
          basePlugins={mcpEditorsPlugin}
          initialValue={parameter.value}
          onChange={onParameterVisibilityUpdate}
          editorBlur={onParameterValueChange}
        />
      );
  }
};
