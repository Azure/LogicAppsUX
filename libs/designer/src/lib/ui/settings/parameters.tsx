import { updateParameterConditionalVisibility } from '../../core/state/operation/operationMetadataSlice';
import type { Settings } from './settingsection';
import type { IDropdownOption } from '@fluentui/react';
import { SearchBox, DropdownMenuItemType, Dropdown } from '@fluentui/react';
import type { FC } from 'react';
import { useState } from 'react';
import { useIntl } from 'react-intl';
import { useDispatch } from 'react-redux';

export interface SearchableParametersDropdownProps {
  conditionallyInvisibleSettings: Settings[];
  groupId: string | undefined;
  nodeId: string;
}

export const SearchableParametersDropdown: FC<SearchableParametersDropdownProps> = ({
  conditionallyInvisibleSettings,
  groupId,
  nodeId,
}): JSX.Element => {
  const showFilterInputItemThreshold = 4;
  const headerKey = 'FilterHeader';

  const intl = useIntl();
  const dispatch = useDispatch();

  const [conditionalVisibilityTempArray, setConditionalVisibilityTempArray] = useState<string[]>([]);
  const [filterText, setFilterText] = useState('');

  const addNewParamText = intl.formatMessage({
    defaultMessage: 'Add new parameters',
    description: 'Text for add new parameter button',
  });

  const searchOperation = intl.formatMessage({
    defaultMessage: 'Search',
    description: 'Placeholder for search box that searches conditional parameters',
  });

  const options = conditionallyInvisibleSettings.map((setting): IDropdownOption => ({
    key: (setting.settingProp as any).id,
    text: (setting.settingProp as any).label ?? '',
  })).filter((option) => option.text.toLowerCase().includes(filterText.toLowerCase()));

  if (conditionallyInvisibleSettings.length >= showFilterInputItemThreshold) {
    options.unshift(
      { key: headerKey, text: '', itemType: DropdownMenuItemType.Header },
      { key: 'FilterDivider', text: '-', itemType: DropdownMenuItemType.Divider },
    );
  }

  return (
    <div style={{ paddingTop: '5px' }}>
      <Dropdown
        className="msla-setting-section-parameter-dropdown"
        placeholder={addNewParamText}
        multiSelect
        options={options}
        selectedKeys={conditionalVisibilityTempArray}
        onChange={(_e: any, item: any) => {
          if (item?.key) {
            setConditionalVisibilityTempArray(
              conditionalVisibilityTempArray.includes(item.key)
                ? conditionalVisibilityTempArray.filter((key) => key !== item.key)
                : [...conditionalVisibilityTempArray, item.key]
            );
          }
        }}
        onDismiss={() => {
          conditionalVisibilityTempArray.forEach((parameterId) => {
            dispatch(updateParameterConditionalVisibility({ nodeId, groupId: groupId ?? '', parameterId, value: true }));
          });
          setConditionalVisibilityTempArray([]);
          setFilterText('');
        }}
        onRenderItem={(item, defaultRenderer) => {
          if (item?.key === headerKey) {
            return (
              <SearchBox
                autoFocus={true}
                className="msla-setting-section-parameter-dropdown-search"
                onChange={(e, newValue) => setFilterText(newValue ?? '')}
                placeholder={searchOperation}
              />
            );
          }

          return defaultRenderer?.(item) ?? null;
        }}
      />
    </div>
  );
};
