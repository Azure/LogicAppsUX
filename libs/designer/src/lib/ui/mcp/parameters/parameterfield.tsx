import type { ChangeState } from '@microsoft/designer-ui';
import { Label, Button, mergeClasses, RadioGroup, Radio, Field, type RadioGroupProps } from '@fluentui/react-components';
import { Dismiss16Regular } from '@fluentui/react-icons';
import { useCallback, useEffect, useMemo, useState } from 'react';
import { useEditOperationStyles } from './styles';
import type { ParameterInfo, ValueSegment } from '@microsoft/logic-apps-shared';
import { useIntl } from 'react-intl';
import { ParameterEditor } from './ParameterEditor';
import constants from '../../../common/constants';
import { parameterHasValue, updateParameterAndDependencies } from '../../../core/utils/parameters/helper';
import { useDispatch, useSelector } from 'react-redux';
import type { AppDispatch, RootState } from '../../../core/state/mcp/store';

type InputType = 'model' | 'user';

export const ParameterField = ({
  operationId,
  groupId,
  parameter,
  isConditional,
  onParameterVisibilityUpdate,
  handleParameterValueChange,
  handleRemoveConditionalParameter,
}: {
  operationId: string;
  groupId: string;
  parameter: ParameterInfo;
  isConditional?: boolean;
  onParameterVisibilityUpdate: () => void;
  handleParameterValueChange: (parameterId: string, newValue: ValueSegment[]) => void;
  handleRemoveConditionalParameter: (parameterId: string) => void;
}) => {
  const intl = useIntl();
  const dispatch = useDispatch<AppDispatch>();
  const styles = useEditOperationStyles();

  const { operationInfo, reference, nodeInputs, dependencies } = useSelector((state: RootState) => ({
    operationInfo: state.operations.operationInfo[operationId],
    reference: state.connection.connectionReferences[state.connection.connectionsMapping[operationId] ?? ''],
    nodeInputs: state.operations.inputParameters[operationId],
    dependencies: state.operations.dependencies[operationId],
  }));

  const [inputType, setInputType] = useState<InputType>(parameterHasValue(parameter) ? 'model' : 'user');

  useEffect(() => {
    setInputType(parameterHasValue(parameter) ? 'model' : 'user');
  }, [parameter]);

  const INTL_TEXT = {
    removeParamText: intl.formatMessage({
      id: '5E66mK',
      defaultMessage: 'Remove parameter',
      description: 'Tooltip for remove parameter button',
    }),
    modelRadioText: intl.formatMessage({
      id: 'LNA+DZ',
      defaultMessage: 'Model',
      description: 'Label for parameter to use model input type',
    }),
    userRadioText: intl.formatMessage({
      id: 'gOWYv9',
      defaultMessage: 'User',
      description: 'Label for parameter to use user input type',
    }),
    inputLabelText: intl.formatMessage({
      id: 'gwYMqA',
      defaultMessage: 'Input',
      description: 'Label for input field in parameter editor radio button section',
    }),
  };

  const isLargeParameter = useMemo(() => {
    const editor = parameter.editor?.toLowerCase();
    return (
      editor === constants.EDITOR.ARRAY ||
      editor === constants.EDITOR.DICTIONARY ||
      editor === constants.EDITOR.HTML ||
      editor === constants.EDITOR.TABLE
    );
  }, [parameter.editor]);

  const onParameterValueChange = useCallback(
    (newState: ChangeState) => {
      const { value: newValueSegments, viewModel } = newState;
      const propertiesToUpdate = { value: newValueSegments, preservedValue: undefined } as Partial<ParameterInfo>;

      if (viewModel !== undefined) {
        propertiesToUpdate.editorViewModel = viewModel;
      }

      dispatch(
        updateParameterAndDependencies({
          nodeId: operationId,
          groupId,
          parameterId: parameter?.id as string,
          properties: propertiesToUpdate,
          isTrigger: false,
          operationInfo,
          connectionReference: reference,
          nodeInputs,
          dependencies,
          updateTokenMetadata: false,
        })
      );
      onParameterVisibilityUpdate();
      handleParameterValueChange(parameter.id, newValueSegments);
    },
    [
      dispatch,
      operationId,
      groupId,
      parameter.id,
      operationInfo,
      reference,
      nodeInputs,
      dependencies,
      onParameterVisibilityUpdate,
      handleParameterValueChange,
    ]
  );

  const handleSelectionChange: RadioGroupProps['onChange'] = (_, data) => {
    setInputType(data.value as InputType);
    // Reset parameter value when model is selected
    if (data.value === 'model') {
      onParameterValueChange({ value: [], viewModel: { hideErrorMessage: true } });
    }
  };

  return (
    <div className={styles.parameterField}>
      <Label className={styles.parameterLabel} required={parameter.required} title={parameter.label}>
        {parameter.label}
      </Label>
      <Field label={INTL_TEXT.inputLabelText} hint={parameter.placeholder}>
        <RadioGroup layout="horizontal" value={inputType} onChange={handleSelectionChange}>
          <Radio value={'model'} key={'model'} label={INTL_TEXT.modelRadioText} />
          <Radio value={'user'} key={'user'} label={INTL_TEXT.userRadioText} />
        </RadioGroup>
      </Field>
      <div className={mergeClasses(styles.parameterValueSection, isLargeParameter && styles.largeParameterSection)}>
        <ParameterEditor
          operationId={operationId}
          groupId={groupId}
          parameter={parameter}
          onParameterVisibilityUpdate={onParameterVisibilityUpdate}
          onParameterValueChange={onParameterValueChange}
        />
        {isConditional && (
          <Button
            appearance="subtle"
            size="small"
            icon={<Dismiss16Regular />}
            onClick={() => handleRemoveConditionalParameter(parameter.id)}
            title={INTL_TEXT.removeParamText}
            className={styles.removeParameterButton}
          />
        )}
      </div>
    </div>
  );
};
