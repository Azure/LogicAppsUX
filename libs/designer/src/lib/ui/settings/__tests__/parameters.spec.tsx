import { store } from '../../../core/store';
import type { SearchableParametersDropdownProps } from '../parameters';
import { SearchableParametersDropdown } from '../parameters';
import type { SettingTextFieldProps } from '@microsoft/designer-ui';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';

jest.mock('@fluentui/react', () => ({
  ...jest.requireActual('@fluentui/react'),
  Dropdown: 'Dropdown',
}));

describe('ui/settings/parameters', () => {
  it('should create a dropdown without a search box for 3 parameters', () => {
    const props: SearchableParametersDropdownProps = {
      conditionallyInvisibleSettings: [
        { settingProp: { id: 'foo', label: 'Foo' } as SettingTextFieldProps, settingType: 'SettingTextField' },
        { settingProp: { id: 'bar', label: 'Bar' } as SettingTextFieldProps, settingType: 'SettingTextField' },
        { settingProp: { id: 'baz', label: 'Baz' } as SettingTextFieldProps, settingType: 'SettingTextField' },
      ],
      groupId: 'group1',
      nodeId: 'Send_an_email',
    };
    const component = (
      <Provider store={store}>
        <SearchableParametersDropdown {...props} />
      </Provider>
    );
    const renderedComponent = renderer.create(component).toJSON();
    expect(renderedComponent).toMatchSnapshot();
  });

  it('should create a dropdown with a search box for 5 parameters', () => {
    const props: SearchableParametersDropdownProps = {
      conditionallyInvisibleSettings: [
        { settingProp: { id: 'foo', label: 'Foo' } as SettingTextFieldProps, settingType: 'SettingTextField' },
        { settingProp: { id: 'bar', label: 'Bar' } as SettingTextFieldProps, settingType: 'SettingTextField' },
        { settingProp: { id: 'baz', label: 'Baz' } as SettingTextFieldProps, settingType: 'SettingTextField' },
        { settingProp: { id: 'qux', label: 'Qux' } as SettingTextFieldProps, settingType: 'SettingTextField' },
        { settingProp: { id: 'qoz', label: 'Qoz' } as SettingTextFieldProps, settingType: 'SettingTextField' },
      ],
      groupId: 'group2',
      nodeId: 'Send_an_email',
    };
    const component = (
      <Provider store={store}>
        <SearchableParametersDropdown {...props} />
      </Provider>
    );
    const renderedComponent = renderer.create(component).toJSON();
    expect(renderedComponent).toMatchSnapshot();
  });
});
