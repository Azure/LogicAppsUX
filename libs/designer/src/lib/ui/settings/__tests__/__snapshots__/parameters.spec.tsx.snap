// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ui/settings/parameters should create a dropdown with a search box for 5 parameters 1`] = `
<div
  style={
    Object {
      "paddingTop": "5px",
    }
  }
>
  <Dropdown
    className="msla-setting-section-parameter-dropdown"
    multiSelect={true}
    onChange={[Function]}
    onDismiss={[Function]}
    onRenderItem={[Function]}
    options={
      Array [
        Object {
          "itemType": 2,
          "key": "FilterHeader",
          "text": "",
        },
        Object {
          "itemType": 1,
          "key": "FilterDivider",
          "text": "-",
        },
        Object {
          "key": "foo",
          "text": "Foo",
        },
        Object {
          "key": "bar",
          "text": "Bar",
        },
        Object {
          "key": "baz",
          "text": "Baz",
        },
        Object {
          "key": "qux",
          "text": "Qux",
        },
        Object {
          "key": "qoz",
          "text": "Qoz",
        },
      ]
    }
    placeholder="Add new parameters"
    selectedKeys={Array []}
  />
</div>
`;

exports[`ui/settings/parameters should create a dropdown without a search box for 3 parameters 1`] = `
<div
  style={
    Object {
      "paddingTop": "5px",
    }
  }
>
  <Dropdown
    className="msla-setting-section-parameter-dropdown"
    multiSelect={true}
    onChange={[Function]}
    onDismiss={[Function]}
    onRenderItem={[Function]}
    options={
      Array [
        Object {
          "key": "foo",
          "text": "Foo",
        },
        Object {
          "key": "bar",
          "text": "Bar",
        },
        Object {
          "key": "baz",
          "text": "Baz",
        },
      ]
    }
    placeholder="Add new parameters"
    selectedKeys={Array []}
  />
</div>
`;
